File: implementations/rust/src/possible_nums.rs
use crate::number::Number;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub struct PossibleNums {
    // using lowest-order 9 bits for numbers 1-9
    bits: u16,
}

impl PossibleNums {
    /// `PossibleNums` with all numbers possible
    pub const ALL: PossibleNums = PossibleNums { bits: 0b111111111 };

    pub fn remove(&mut self, num: Number) {
        self.bits &= !(1 << (num.to_u8() - 1));
    }

    pub fn single(&self) -> Option<Number> {
        if self.bits.count_ones() == 1 {
            let mut bits = self.bits;
            let mut index = 0;
            while bits != 0 {
                if bits & 1 == 1 {
                    return Some(Number::from_u8_normalize(index + 1));
                }
                bits >>= 1;
                index += 1;
            }
        }
        None
    }

    pub fn iter(&self) -> PossibleNumsIterator {
        PossibleNumsIterator {
            bits: self.bits,
            index: 0,
        }
    }
}

impl IntoIterator for PossibleNums {
    type Item = Number;
    type IntoIter = PossibleNumsIterator;

    fn into_iter(self) -> Self::IntoIter {
        PossibleNumsIterator {
            bits: self.bits,
            index: 0,
        }
    }
}

impl IntoIterator for &PossibleNums {
    type Item = Number;
    type IntoIter = PossibleNumsIterator;

    fn into_iter(self) -> Self::IntoIter {
        self.iter()
    }
}

pub struct PossibleNumsIterator {
    bits: u16,
    index: u8,
}

impl Iterator for PossibleNumsIterator {
    type Item = Number;

    fn next(&mut self) -> Option<Self::Item> {
        while self.index < 9 {
            let mask = 1 << self.index;
            self.index += 1;
            if self.bits & mask != 0 {
                return Some(Number::from_u8_normalize(self.index));
            }
        }
        None
    }
}

#[cfg(test)]
mod tests {
    use crate::possible_nums::PossibleNumsIterator;

    use super::Number;
    use super::PossibleNums;

    #[test]
    fn remove() {
        let mut possible_nums = PossibleNums::ALL;

        possible_nums.remove(Number::One);
        assert_eq!(possible_nums.bits, 0b111111110);

        possible_nums.remove(Number::Nine);
        assert_eq!(possible_nums.bits, 0b011111110);
    }

    #[test]
    fn single() {
        let possible_nums = PossibleNums { bits: 0b000000010 };
        assert_eq!(possible_nums.single(), Some(Number::Two));

        let possible_nums = PossibleNums { bits: 0b000000000 };
        assert_eq!(possible_nums.single(), None);

        let possible_nums = PossibleNums { bits: 0b001000010 };
        assert_eq!(possible_nums.single(), None);
    }

    #[test]
    fn iterator() {
        let mut iter = PossibleNumsIterator {
            bits: 0b100111111,
            index: 0,
        };

        assert_eq!(iter.next(), Some(Number::One));
        assert_eq!(iter.next(), Some(Number::Two));
        assert_eq!(iter.next(), Some(Number::Three));
        assert_eq!(iter.next(), Some(Number::Four));
        assert_eq!(iter.next(), Some(Number::Five));
        assert_eq!(iter.next(), Some(Number::Six));
        assert_eq!(iter.next(), Some(Number::Nine));
        assert_eq!(iter.next(), None);
    }

    #[test]
    fn into_iter() {
        let possible_nums = PossibleNums { bits: 0b100000010 };

        let mut iter = possible_nums.into_iter();

        assert_eq!(iter.next(), Some(Number::Two));
        assert_eq!(iter.next(), Some(Number::Nine));
        assert_eq!(iter.next(), None);
    }

    #[test]
    fn iter() {
        let possible_nums = PossibleNums { bits: 0b100000010 };

        let mut iter = possible_nums.iter();

        assert_eq!(iter.next(), Some(Number::Two));
        assert_eq!(iter.next(), Some(Number::Nine));
        assert_eq!(iter.next(), None);
        assert_eq!(possible_nums.bits, 0b100000010)
    }
}

File: implementations/zig/src/possible_nums.zig
const std = @import("std");
const BitSet = std.bit_set.IntegerBitSet(9);
const Number = @import("number.zig").Number;

pub const PossibleNums = struct {
    const Self = @This();
    bits: BitSet,

    pub const INIT_ALL: PossibleNums = Self{ .bits = BitSet.initFull() };

    pub fn remove(self: *Self, num: Number) void {
        self.bits.unset(num.toInt() - 1);
    }

    pub fn single(self: Self) ?Number {
        if (self.bits.count() == 1) {
            if (self.bits.findFirstSet()) |i| {
                return Number.fromIntClamp(i + 1);
            }
        }

        return null;
    }
    pub const List = std.BoundedArray(Number, 9);
    pub fn list(self: Self) List {
        var nums = List.init(0) catch unreachable;

        var iter = self.iterator();
        while (iter.next()) |num| {
            nums.append(num) catch unreachable;
        }

        return nums;
    }

    pub fn iterator(self: Self) Iterator {
        return Iterator{ .bitIter = self.bits.iterator(.{}) };
    }

    pub const Iterator = struct {
        const IterSelf = @This();

        bitIter: BitSet.Iterator(.{}),

        pub fn next(self: *IterSelf) ?Number {
            if (self.bitIter.next()) |i| {
                return Number.fromIntClamp(i + 1);
            } else {
                return null;
            }
        }
    };
};

test PossibleNums {
    const expectEqual = std.testing.expectEqual;
    var possible = PossibleNums.INIT_ALL;
    possible.remove(Number.two);
    try expectEqual(null, possible.single());

    var iter = possible.iterator();
    try expectEqual(Number.one, iter.next());
    try expectEqual(Number.three, iter.next());

    possible.remove(Number.one);
    possible.remove(Number.three);
    possible.remove(Number.four);
    possible.remove(Number.six);
    possible.remove(Number.seven);
    possible.remove(Number.eight);
    possible.remove(Number.nine);

    try expectEqual(Number.five, possible.single());
}

// test "PossibleNums.Iterator" {
//     var possible = PossibleNums.ALL;
//     var iter1 = possible.iterator();

//     std.debug.print("\nAll Numbers:\n", .{});
//     while (iter1.next()) |num| {
//         std.debug.print("{any}\n", .{num});
//     }

//     possible.remove(Number.one);
//     possible.remove(Number.five);
//     possible.remove(Number.nine);

//     var iter2 = possible.iterator();
//     std.debug.print("\nFewer Numbers:\n", .{});
//     while (iter2.next()) |num| {
//         std.debug.print("{any}\n", .{num});
//     }

//     std.debug.print("\nA Single Number?\n", .{});
//     if (possible.single()) |num| {
//         std.debug.print("Yep: {any}", .{num});
//     } else {
//         std.debug.print("Nope.", .{});
//     }

//     possible.remove(Number.two);
//     possible.remove(Number.three);
//     possible.remove(Number.four);
//     possible.remove(Number.six);
//     possible.remove(Number.eight);

//     std.debug.print("\nA Single Number?\n", .{});
//     if (possible.single()) |num| {
//         std.debug.print("Yep: {any}", .{num});
//     } else {
//         std.debug.print("Nope.", .{});
//     }

//     std.debug.print("\ndone\n", .{});
// }
